addons/stock/models/product.py

```diff
@@ -759,6 +759,18 @@ def _onchange_type(self):
         return res
 
     def write(self, vals):
+        if 'company_id' in vals and vals['company_id']:
+            for product in self:
+                if product.company_id.id == vals['company_id']:
+                    continue
+            # Forbids to change the product's company it it exists quants in other company.
+            quants = self.env['stock.quant'].search([
+                ('product_id', 'in', product.product_variant_ids.ids),
+                ('company_id', '!=', vals['company_id']),
+            ])
```

Comment:
Do we need to worry about the case of zero quants blocking the change? Since `_quant_tasks`/`_unlink_zero_quants` is only called at specific times (i.e. scheduler + opening specific views). I.e. should we run `_unlink_zero_quants` before doing the search? Or add more to the domain to avoid the quants that shouldn't block the change?

Also, could we improve this by adding a limit=1 (i.e. so the search ends as soon as it finds the first one)?

Reply:
I think we can ignore quants with 0 quantity indeed :) (I added that in the domain).
About the `limit`, I don't think it will work like that, so instead I'll use `count` as it makes the query simpler

Reply:
If you add `limit=1`, it will act as the SQL query command LIMIT=1 which should be run auto-magically optimized to stop the search once the first row matching the domain is found, unless there's something unique about the db setup that I don't know :thinking:

Reply:
Sorry I didn't see you responded before today :grimacing: 
I think it will do that only if there is no order, otherwise, even with a `limit` it needs to find all matching records to returns only the N first ones matching the domain after sorted them (N depending of the given limit).
This is why I wanted to use `count` over `limit`, but it's also true we can force no order so I'm also OK with that :)
(I think there is no order on `stock.quant` anyway but I prefer to force it to be sure)

Also, in the meantime I found out it the user hasn't access to one of the quants' company, the quant will not be found by the `search` since it belongs to a company the user has no access.
So I added a `sudo` to find the quants even if they are in a company the user cannot see.